### PR TITLE
Add approx_derive

### PR DESCRIPTION
**Pull request by schmidma**
## Introduced Changes

Implement a new `approx_derive` crate to derive `AbsDiffEq` und `RelativeEq` of the `approx` crate.

I'm open to discussion whether this change is actually useful...

## ToDo / Known Issues

none

## Ideas for Next Iterations (Not This PR)

none

## How to Test

compile?

### CODE REVIEW

File: crates/approx_derive/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                let epsilon = string
+                    .parse()
+                    .expect_or_abort("failed to parse epsilon type");
+                Some(epsilon)
+            }
+            _ => None,
+        })
+}
+
+fn parse_meta_items(attribute: &Attribute) -> Option<punctuated::IntoIter<NestedMeta>> {
```
Comment (knoellle) [inline crates/approx_derive/src/lib.rs:79]: This parses just one item, doesn't it?
```suggestion
fn parse_meta_item(attribute: &Attribute) -> Option<punctuated::IntoIter<NestedMeta>> {
```

File: crates/approx_derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                let string = match lit {
+                    Lit::Str(string) => string,
+                    _ => abort!(lit, "expected string literal"),
+                };
+                let epsilon = string
+                    .parse()
+                    .expect_or_abort("failed to parse epsilon type");
+                Some(epsilon)
+            }
+            _ => None,
```
Comment (knoellle) [inline crates/approx_derive/src/lib.rs:75]: Shouldn't an unknown path name in an `abs_diff_eq` attribute cause an error?

File: crates/types/src/ball.rs
Comment (knoellle) [inline crates/types/src/ball.rs:1]: Is there a particular reason you only applied this here but not to all the other types with manual implementations for these crates?

File: crates/approx_derive/src/lib.rs (hunk 1 above)
Comment (schmidma) [reply crates/approx_derive/src/lib.rs:79]: hmmm... :thinking: there is one `Meta`, but several items in that meta... And the function is returning an iterator (-of multiple items)

File: crates/approx_derive/src/lib.rs (hunk 2 above)
Comment (schmidma) [reply crates/approx_derive/src/lib.rs:75]: yep, i think this is correct, but this function should find an `epsilon` definition, not knowing if any other definitions are present. I would have to implement a parallel check for allowed arguments in the attribute list. Do you think this is worth it?

File: crates/types/src/ball.rs
Comment (schmidma) [reply crates/types/src/ball.rs:1]: No, just did not want to make the change to big, and also some of the implementations differ from the derivable impl

**Review by knoellle — DISMISSED** [review]
LGTM, except this PR has conflicts.
please rebase

**Review by knoellle — APPROVED** [review]
LGTM

**Pull request merged**